### PR TITLE
[AndroidCrypto] Handle setting non-default SslProtocols

### PR DESCRIPTION
**Pull request by elinor-fung**
- Set enabled `SslProtocols`
  - Throw PNSE if requested combination is not contiguous - matches macOS behaviour. Android actually does allow specifying a non-contiguous set, but it will only respect the lowest contiguous set and ignore the rest, so it seemed more consistent with our APIs to explicitly throw.
  - `Ssl2` and `Ssl3` are not supported, `Tls13` is only supported on some (newer) versions
- Always enlarge buffer when `wrap`/`unwrap` gives `BUFFER_OVERFLOW`
- Fix some missed local ref deletes
- Throw PNSE  if `EncryptionPolicy.NoEncryption` is requested (not supported on Android)
- Update test utilities to properly check for support for `Ssl3` and `Tls13` on Android

cc @jkoritzinsky @steveisok @AaronRobinsonMSFT @bartonjs @wfurt

### CODE REVIEW

File: src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_ssl.c [hunk 1]
```diff
@@ -3,34 +3,55 @@
...
-    for (int i = 0; i < protocolsCount; i++)
+    // SSLContext context = SSLContext.getDefault();
+    // SSLParameters params = context.getDefaultSSLParameters();
+    // String[] protocols = params.getProtocols();
+    loc[context] = (*env)->CallStaticObjectMethod(env, g_sslCtxClass, g_sslCtxGetDefaultMethod);
+    loc[params] = (*env)->CallObjectMethod(env, loc[context], g_sslCtxGetDefaultSslParamsMethod);
+    loc[protocols] = (*env)->CallObjectMethod(env, loc[params], g_SSLParametersGetProtocols);
+
+    const char tlsv1[] = "TLSv1";
+    size_t tlsv1Len = (sizeof(tlsv1) - 1) / sizeof(char);
```
Comment (AaronRobinsonMSFT) [inline src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_ssl.c:20]: The typical method for this is:

```C
size_t tlsv1Len = (sizeof(tlsv1) / sizeof(*tlsv1)) - 1;
```

File: src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_sslstream.c [hunk 2]
```diff
@@ -750,6 +755,53 @@ int32_t AndroidCryptoNative_SSLStreamGetPeerCertificates(SSLStream* sslStream, j
...
+int32_t AndroidCryptoNative_SSLStreamSetEnabledProtocols(SSLStream* sslStream, PAL_SslProtocol* protocols, int32_t count)
+{
+    assert(sslStream != NULL);
+
+    JNIEnv* env = GetJNIEnv();
+    int32_t ret = FAIL;
+
+    // String[] protocolsArray = new String[count];
+    jobjectArray protocolsArray = (*env)->NewObjectArray(env, count, g_String, NULL);
+    for (int i = 0; i < count; ++i)
```
Comment (AaronRobinsonMSFT) [inline src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_sslstream.c:787]: Nit: `int` -> `int32_t`

**Review by jkoritzinsky — APPROVED** [review]
Interop work looks good to me.

**Review by wfurt — APPROVED** [review]
LGTM.
I personally feel it would be ok to not worry about SSL2&3. 
We drag them as legacy compat but they should be long dead. 
Since Android probably does not have that baggage we may have more freedom.
Many Linux distributions do not supported them in practice as well.

**Comment by elinor-fung** [comment]
I like that thought of not dragging along SSL3 as legacy compat - especially since the Android platform dropped support for it a couple of versions/years ago.

**Pull request merged**